gk/fib.c

```diff
@@ -0,0 +1,1343 @@
+
+	if ((gt_prefix_fib->action != GK_FWD_GATEWAY_BACK_NET &&
+			gt_prefix_fib->action != GK_FWD_NEIGHBOR_BACK_NET)) {
+		return NULL;
+	}
+
+	gt_fib = add_prefix_fib(gt_addr, prefix_len, gk_conf);
+	if (gt_fib == NULL)
+		goto out;
+
+	gt_fib->action = GK_FWD_GRANTOR;
+	flow = &gt_fib->u.grantor.flow;
+
+	if (gt_addr->proto == ETHER_TYPE_IPv4) {
+		flow->proto = ETHER_TYPE_IPv4;
+		flow->f.v4.src = gk_conf->net->back.ip4_addr.s_addr;
+		flow->f.v4.dst = gt_addr->ip.v4.s_addr;
+	} else {
+		flow->proto = ETHER_TYPE_IPv6;
+		rte_memcpy(flow->f.v6.src,
+			gk_conf->net->back.ip6_addr.s6_addr,
+			sizeof(flow->f.v6.src));
+		rte_memcpy(flow->f.v6.dst,
+			gt_addr->ip.v6.s6_addr, sizeof(flow->f.v6.dst));
+	}
+
+	/* Fill up the @next_fib field in the @gt_fib. */
+
+	/*
+	 * The Grantor server is not an neighbor.
+	 * So the @next_fib needs to point to the
+	 * responsbile gateway entry.
+	 */
+	if (gt_prefix_fib->action == GK_FWD_GATEWAY_BACK_NET) {
+		gt_fib->u.grantor.next_fib = gt_prefix_fib;
+		gt_fib->u.grantor.next_fib->ref_cnt++;
+		gt_fib->u.grantor.eth_cache =
+			gt_prefix_fib->u.gateway.eth_cache;
+
+		/* TODO Add concurrency control. */
+		gt_fib->u.grantor.eth_cache->ref_cnt++;
+
+		goto out;
+	}
+
+	/* The Grantor server is an neighbor. */
+	if (gt_addr->proto == ETHER_TYPE_IPv4) {
+		eth_cache = lookup_ether_cache(
+			&gt_prefix_fib->u.neigh,
+			&gt_addr->ip.v4.s_addr);
+		if (eth_cache == NULL) {
+			ret = rte_hash_add_key(
+				gt_prefix_fib->u.neigh.hash_table,
+				&gt_addr->ip.v4.s_addr);
+			if (ret >= 0) {
+				eth_cache = &gt_prefix_fib->
+					u.neigh.cache_tbl[ret];
+			} else {
+				RTE_LOG(ERR, HASH,
+					"Failed to add a cache entry to the neighbor hash table at %s\n",
+					__func__);
+				return NULL;
+			}
+		}
+
+	} else {
+		eth_cache = lookup_ether_cache(
+			&gt_prefix_fib->u.neigh6,
+			gt_addr->ip.v6.s6_addr);
+		if (eth_cache == NULL) {
+			ret = rte_hash_add_key(
+				gt_prefix_fib->u.neigh6.hash_table,
+				gt_addr->ip.v6.s6_addr);
+			if (ret >= 0) {
+				eth_cache = &gt_prefix_fib->
+					u.neigh6.cache_tbl[ret];
+			} else {
+				RTE_LOG(ERR, HASH,
+					"Failed to add a cache entry to the neighbor hash table at %s\n",
+					__func__);
+				return NULL;
+			}
+		}
+	}
+
+	eth_cache->stale = true;
+	eth_cache->eth_hdr.ether_type = gt_addr->proto;
+
+	/* Grantor server must be forwarded to the back network. `*/
+	ether_addr_copy(&gk_conf->net->back.eth_addr,
+		&eth_cache->eth_hdr.s_addr);
+
+	eth_cache->ref_cnt++;
+
+	gt_fib->u.grantor.eth_cache = eth_cache;
+	gt_fib->u.grantor.next_fib = gt_prefix_fib;
+	gt_fib->u.grantor.next_fib->ref_cnt++;
+
+out:
+	return gt_fib;
+}
+
+static int
+add_fib_entry(struct lua_gk_fib *gk_fib, struct gk_config *gk_conf)
+{
+	int ret;
+
+	struct ipaddr gw_addr;
+	struct gk_fib *gw_fib = NULL;
+
+	struct ipaddr gt_addr;
+	struct gk_fib *gt_fib = NULL;
+
+	struct ipaddr ip_prefix_addr;
+	int ip_prefix_len;
+	struct gk_fib *ip_prefix_fib = NULL;
+
+	switch (gk_fib->action) {
```

Comment:
Ok, I commented on the wrong spot and further reading the code I saw that there's an advantage with combining IPv4 and IPv6 in the same call, but I still think that we can simplify the code. Having the following functions exported to Lua, instead of the current interface, would simplify the code in C and in Lua:

add_fib_entry(const char *ip_prefix, enum gk_fib_action action, const char *grantor_or_gateway);
del_fib_entry(const char *ip_prefix);

These functions would eliminate the need for `enum gk_fib_op`, `struct lua_gk_fib`, `dy_gk_fib_proc()`, `lua_gk_fib_proc()`, and Lua code would, instead of creating tables like `fib_entries`, just call those functions.